src/wallet/mod.rs

```diff
@@ -253,32 +261,91 @@ where
             .map_err(|_| Error::ScriptDoesntHaveAddressForm)
     }
 
-    // Return the the last previously derived address for `keychain` if it has not been used in a
-    // received transaction. Otherwise return a new address using [`Wallet::get_new_address`].
-    fn get_unused_address(&self, keychain: KeychainKind) -> Result<AddressInfo, Error> {
+    // Return vector of unused address indexes
+    fn get_unused_key_indexes(&self, keychain: KeychainKind) -> Result<Vec<u32>, Error> {
         let current_index = self.fetch_index(keychain)?;
 
-        let derived_key = self
-            .get_descriptor_for_keychain(keychain)
-            .as_derived(current_index, &self.secp);
-
-        let script_pubkey = derived_key.script_pubkey();
+        let derived_keys: Vec<Descriptor<DerivedDescriptorKey>> = (0..current_index + 1)
+            .map(|i| {
+                self.get_descriptor_for_keychain(keychain)
+                    .as_derived(i, &self.secp)
+            })
+            .collect();
```

Comment:
This should be using the scripts (`iter_script_pubkeys`) in the database not re-deriving them which is costly.